### PR TITLE
feat(polls) Ability to create polls inside Jitsi

### PR DESCRIPTION
**Pull request by jade-guiton**
# Description

This PR merges a feature developed during the "Winning over the classroom with Jitsi" hackathon.

## Features and current design

- "Polls" tab in chat sidebar
- "Create a poll" button in tab
- Poll creation form in tab
  - Add/delete answer fields (+ reorder on web version)
- Display of current polls in sidebar
  - Select multiple answers or skip to answers
  - Revert answer
  - View "condensed" results (percentages) or "details" (names)
- Web + Native support
- English and French translation support

## Note on poll persistence

We use a Prosody module to allow new participants who join a room to receive polls that were previously sent to that room. In order to simplify the implementation, we use a new feature of lib-jitsi-meet to handle client-client and client-server communication. The feature was added recently [in this Pull Request](https://github.com/jitsi/lib-jitsi-meet/pull/1671). Thus, the version of lib-jitsi-meet used in jitsi-meet may have to be updated before this PR can be merged.

## For posterity: original hackathon design

- "Create a poll" button in toolbar
- Poll creation dialog:
  - Add/reorder/delete answer fields
- Poll answer dialog:
  - Multiple choice by default
  - Ability to skip answering
  - Show results in dialog after answering
- Poll results in chat after answering:
  - "Details" button show individual votes

Screenshots and a video demonstration of the hackathon version can be found on [the hackathon project page](https://platform-euhack21.bemyapp.com/#/projects/6082a2402f90ac001965e2bf).

- Bugs fixed after hackathon:
  - ~UI overflow when questions/answers are too long (web + native)~ (fixed in commit fb74869)
  - ~Names in poll results change to "Absent Jitster" when voter leaves~ (fixed in commit c6cd7ed)
  - ~Dialog close buttons do not work properly~ (fixed in commit 41175fa)
  - ~Poll answer modals open "spontaneously", so they can "overwrite" other modals~ (fixed in commit a7bbc12)

### CODE REVIEW

**Comment by spineki** [comment]
## Description of the poll communication protocol:

## Current poll communication protocol:

When a new poll is created, a message is broadcast to all users containing the following information:
a randomly generated poll ID
- the ID of the author of the poll
- the question
- the possible answers

When a user answers a poll, another message is broadcast containing the following information:
- the poll ID
- the user ID
- an array of booleans, matching the list of possible answers

Currently, this data is processed client-side, by each client independently. We would like to add some persistence to poll results, so that newcomers to a room can see previous polls, and potentially answer them. Before doing so, however, we would like to make sure we are using the best method for implementing this protocol.

## Implementations:

### Current implementation:

Both kinds of messages are sent through the sendCommandOnce method on the lib-jitsi-meet conference object. Each type of message has a command name associated with it (COMMAND_NEW_POLL and COMMAND_ANSWER_POLL). The message contents are first turned into an XML tree so they can be sent through the XMPP MUC channel.

This method uses a “one-time presence” stanza to transmit information: the function adds an XMPP presence object to the user, with the given data, then sends the presence update, then removes the presence, without sending an update.
Because presence data is automatically broadcast to all users, there was no need for modifications on the backend.

This works, but has 2 major drawbacks:

- Adding persistence to this system is complicated, as it implies fishing out the poll data from the presence stanzas on the backend, then reemitting such a stanza when a new user joins the room, without disturbing other parts of Jitsi that use presence.
- The sendCommand is kind of a hack, exploiting the presence mechanism in a way that XMPP was not designed for, and it would be good to try to use something cleaner.
 
### Proposals for improvements:

1. We could send custom XMPP messages using Strophe.js (the library used by lib-jitsi-meet to interface with XMPP) directly. We would then add backend logic to broadcast and persist those messages manually.
This could be done in two ways:
    - Directly from jitsi-meet: this avoids making changes to lib-jitsi-meet, but this is somewhat contrary to the high level of abstraction intended for jitsi-meet, and would add a new dependency to it, as the version of Strophe.js used by lib-jitsi-meet does not seem to be readily accessible from jitsi-meet.
    - In lib-jitsi-meet: this would require adding new functions to the lib-jitsi-meet API specifically for sending and receiving polls.
    - However, this could also be made into a generic “data sending” function. The problem is that we would now have 3 different ways of sending data between client and server…
alt text
It would also be possible to replace the way sendCommandOnce currently works by this, but this might be out of the scope of our PR.  
2. We could use the “sendMessage” method on the lib-jitsi-meet conference object to send “endpoint messages” / “JSON messages” to the XMPP server, the same way the SpeakerStats module does.

Again, there seem to be two methods:
- We could make use of the automatic broadcasting of such messages by the MUC module to avoid having to reimplement the broadcasting system, record the messages as they pass through, then send them again when a user joins.
However, the current client logic for receiving JSON messages prevents receiving messages sent by oneself, or sent in a delayed fashion when joining like for chat messages.   
This could be solved by either:   
    - adding special logic for receiving one’s own messages and somehow sending the join-time messages some other way, which is a bit messy;
    - or modifying the current client reception logic (with the risk of breaking other features that use this communication method).
- We could copy the way SpeakerStats works, only send messages directly between the client and the server, and broadcast messages manually from the server.
However, this seems like a waste if the messages could be broadcast automatically.

**Comment by damencho** [comment]
Do you prefer discussing here or in the community forum? 

I was just looking at sendComand stuff and how we use it in other places... and yeah probably we want to move away from using presence for it (that is different topic and away from the scope of this PR, which I will start today wth the other team members).

> Use of JitsiConference.sendCommandOnce to create / answer polls
> How best to add persistence (show past polls to newcomers, like chat)

Why are you using sendCommandOnce? If all participants use sendCommand the value will be stored in their presence and in the room state and once a new participant joins he/she will receive the presences from all participants including those commands. And the one joining can recreate the state, just mind that participant can receive answers first then the poll creation. If you use sendCommand, you need to take care that you can receive the data multiple times :)

We use presence for youtube-video sharing, for opening etherpad, local recording, recorder-info???, follow-me, (for avatar and email info which needs to stay in presence I think ... will figure this out).  (Some of these are notes to myself:) )

I would say, use sendCommand, make it work and the rest is something we will need to handle in ljm and as prosody modules :) 
Seems like creating a generic way of doing that stuff ... is beyond the work in this PR. I know we started a conversation in another direction while talking offline ... but now looking at stuff I think this is the best way going forward.

**Comment by jade-guiton** [comment]
> I would say, use sendCommand, make it work and the rest is something we will need to handle in ljm and as prosody modules

The problem with using sendCommand is that it does not allow for real persistence if the author of the poll leaves the room.

It might also be a bit complicated to allow making multiple polls per person, though that can be solved.

Another problem I see with it is that all of the information (questions, answers, and votes) for every poll would be sent again for every presence update, which might be a lot of data, that would end up being ignored.

Speaking of Prosody modules, what do you think of the proposal in commit b98cd8b ? It seems to work fine on our test server (you just need to enable the "polls" module under the "muc" component in the conference subdomain).

**Comment by damencho** [comment]
> The problem with using sendCommand is that it does not allow for real persistence, since if the author of the poll leaves the room, I think the information would be lost. It might also be a bit complicated to allow making multiple polls per person, though that can be worked around.

I agree.



> Another problem I see with it is that all of the information (questions, answers, and votes) for every poll would be sent again for every presence update, which might be a decent amount of data, that would end up being ignored.

I agree, that's why we want to move away from it :)



> Speaking of Prosody modules, what do you think of the proposal in commit [b98cd8b](https://github.com/jitsi/jitsi-meet/commit/b98cd8b41c0711eacf38ed25d00236ec7f5c3a77) ? It seems to work fine on our test server (you just need to enable the "polls" module under the "muc" component in the conference subdomain).

Looks good, maybe something like the generic one, how it needs to be ... :) So I like it, but need to get a deeper look, but in general yes 👍 

There is one thing `conference.room.addListener('xmmp.json_message_received', (senderJid, data) => { `

Is conference JitsiConference? 
We should not access internals (like .room if that is ChatRoom) of lib-jitsi-meet, and we should not be listening to internal events like 'xmmp.json_message_received'.

You can just listen for JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED on conference. 
You will answer me: 'but we do not receive our own messages'. And here is my question :): why do you need that, what information you are expecting to receive that you don't already have?

**Comment by jade-guiton** [comment]
> Is conference JitsiConference?
> We should not access internals (like .room if that is ChatRoom) of lib-jitsi-meet, and we should not be listening to internal events like 'xmmp.json_message_received'.
> You can just listen for JitsiConferenceEvents.ENDPOINT_MESSAGE_RECEIVED on conference.
> You will answer me: 'but we do not receive our own messages'. And here is my question :): why do you need that, what information you are expecting to receive that you don't already have?

We don't really *need* that. There are multiple options that would allow us to avoid using this internal event:
1. Add special logic to handle local polls differently from remote polls
  (makes the code more complex, and a bit of a waste since the message *is* being broadcast back to the user, we just don't get to see it, all because of some specific logic that was probably meant for the subtitles feature originally)
2. Modify lib-jitsi-meet so that ENDPOINT_MESSAGE_RECEIVED doesn't filter local messages
  (might require changing some code for features that already use this event)
  In my opinion, this option is nice because it would basically give us a ready-made and general-purpose replacement for sendCommandOnce.
3. Modify lib-jitsi-meet to add another event that doesn't filter local messages, or expose JSON_MESSAGE_RECEIVED to the outside
  (a bit of a waste since the two events will be almost identical)

What option do you think would be best?

**Comment by damencho** [comment]
I would say 1 ... but I get your point about the code ... 

But at the end of the day isn't it all a matter of calling one method.

```
conference.on('endpoiint_message', process_incomfingMessage);

--- create a poll
let newMsg = {some_json: ...};
sendMessage(newMsg);
// if needed newMsg = transform(newMsg);
process_incomingMessage(newMsg);
```

You get my point, right? There should be way to simplify the code and minimize the duplications and processing ...

**Comment by damencho** [comment]
One question,if the prosody module is not updated, what will happen? Will the UI detect it and hide all polls buttons and stuff.

**Comment by jade-guiton** [comment]
> One question,if the prosody module is not updated, what will happen? Will the UI detect it and hide all polls buttons and stuff.

The feature should continue to work properly, just without persistence. That is, users that join a call will not be able to see polls created before they joined.

**Comment by jade-guiton** [comment]
> I would say 1 ... but I get your point about the code ...
> But at the end of the day isn't it all a matter of calling one method.
> You get my point, right? There should be way to simplify the code and minimize the duplications and processing ...

I hadn't thought of it that way. We'll try to make that change tomorrow.

**Comment by jade-guiton** [comment]
After thinking about it a bit more and testing it, I realized there is another problem with using ENDPOINT_MESSAGE_RECEIVED: there is no way to receive a message directly from the backend with it. The reason why it filters out local messages is that it attempts to retrieve the JitsiParticipant object corresponding to the sender for each message, and aborts if it doesn't exist. This means that it also filters out messages that come directly from Prosody rather than a participant.

It would theoretically be possible to send each poll separately, and assign the "from" based on the original creator of the poll — this is what is done for historic chat messages — but this won't work if that user is no longer in the call.

What do you suggest @damencho?

**Comment by damencho** [comment]
From is not something you can play with ... So your observations lead me to think, that there is no way of getting out of this without changing ljm ... But I wonder what is the best way to do it ... Whether to modify current endpoint message event passing null participant and a boolean at the end saying issystem=true ... but is there a way to accommodate and your own messages received .... Or add a new event with different type sys_endpoint_message or better name ... I'm in doubt 
@saghul do you have a preference?

**Comment by sampaccoud** [comment]
We deployed this feature to our Jitsi platform in production and got some feedback:
- if a user joins the call after a poll is launched, s.he will see the results with the button "change vote" although s.he did not vote yet.
- if a user reloads the page and chooses to change vote, the vote will be counted as a new one. I understand that a user will always be able to vote several times by lauching several browsers or clearing the browser session... but at least we could recognize a user who just reloads the page or switches to a new tab, the same way as Jitsi remembers our backgrounds and usernames?

**Comment by jade-guiton** [comment]
For the first point:

> * if a user joins the call after a poll is launched, s.he will see the results with the button "change vote" although s.he did not vote yet.

The reason we decided to default to the results display is that we assumed most people who join a call late would not want/need to vote on "old" polls, but might still be interested in the results.

- If that is fine and the problem is just the name of the button, we could rename it to "Vote" when no vote has been cast (which would also be more accurate if the user clicked on "Skip to results").

- The other option would be to default to the answering form instead of the results display. Do you think that would be a better idea?

**Comment by sampaccoud** [comment]
> For the first point:
> 
> > * if a user joins the call after a poll is launched, s.he will see the results with the button "change vote" although s.he did not vote yet.
> 
> The reason we decided to default to the results display is that we assumed most people who join a call late would not want/need to vote on "old" polls, but might still be interested in the results.
> 
>     * If this behavior is fine and the problem is just the name of the button, we could simply rename it to "Vote" when no vote has been cast (which would also be more accurate if the user clicked on "Skip to results").
> 
>     * The other option would be to default to the answering form instead of the results display. Do you think that would be a better idea?

Yes! I was just referring to the text of the button. The fact that the user arrives on the results is fine with us.

**Comment by sampaccoud** [comment]
Continuing our tests : it seems that when a user changes his/her vote, results can go above 100%. For example here 106%:

**Comment by jade-guiton** [comment]
> Continuing our tests : it seems that when a user changes his/her vote, results can go above 100%. For example here 106%:

This is the intended behavior. Each answer shows the percentage of respondants who voted for that option, not the percentage of total votes.

So, if someone votes for multiple answers, the total will add up to more than 100%. This is the same behavior as in Telegram multi-option polls for example.

Example : for a poll with options "1" and "2", and respondents Alice and Bob, if "1" is chosen by Alice and Bob, and "2" is chosen by Alice, then "1" will show 100% (100% of all voters voted for "1"), and "2" will show 50% (1/2 voters voted for "2"). The total here is 150%.

> Another small issue: the links "show/hide details" are inverted.

Oops, that's a translation error on our part. We will add it to the list.

**Comment by sampaccoud** [comment]
> > Continuing our tests : it seems that when a user changes his/her vote, results can go above 100%. For example here 106%:
> 
> This is the intended behavior. Each answer shows the percentage of respondants who voted for that option, not the percentage of total votes.

Ok I understand. But in this case, would it be clearer to see my earlier vote when I click on "change vote"?

**Comment by jade-guiton** [comment]
> Ok I understand. But in this case, would it be clearer to see my earlier vote when I click on "change vote"?

You mean, pre-filling the answer form with the previously chosen answers when clicking on "change vote"? That sounds like a good idea. I'll add it to the backlog as well.

If the issue is just seeing what you voted for, it would also be possible to add indicators to the results display, like little check marks to see what you personally voted for.

**Comment by sampaccoud** [comment]
> > Ok I understand. But in this case, would it be clearer to see my earlier vote when I click on "change vote"?
> 
> You mean, pre-filling the answer form with the previously chosen answers when clicking on "change vote"? That sounds like a good idea. I'll add it to the backlog as well.
> 
> If the issue is just seeing what you voted for, it would also be possible to add indicators to the results display, like little check marks to see what you personally voted for.

Yes I meant the first one. And you second point is a good idea but more of a "cherry on the cake" :wink:

**Comment by jade-guiton** [comment]
FYI, I've created 2 PRs over on the fork repository to make the UI changes we discussed above:
- fixing the "show/hide details" buttons
- improving "change vote" behavior ("vote" vs. "change vote" + fill in previous answers in answer form).

We will (hopefully) review and merge them into the "polls" branch soon.

**Comment by damencho** [comment]
> The reason we decided to default to the results display is that we assumed most people who join a call late would not want/need to vote on "old" polls, but might still be interested in the results.
> 
> * If this behavior is fine and the problem is just the name of the button, we could simply rename it to "Vote" when no vote has been cast (which would also be more accurate if the user clicked on "Skip to results").
> * The other option would be to default to the answering form instead of the results display. Do you think that would be a better idea?

Defaulting to the voting part is also ok. Or changing the button text from `Change vote` to `Vote` is also fine. 
This behavior confused me while I did the first tests ... maybe the confusion came more from the text of the button, as I haven't voted and I see a change vote.

**Comment by j-niemczyk** [comment]
I think it would be a big help for many people if there was an option to export the results to even a simple .txt file. Would it be possible to add such an option?

**Review by damencho — CHANGES_REQUESTED** [review]
Small improvement: can you add some docs to the mod_polls.lua code before every method describing what it does, thank you.

**Comment by damencho** [comment]
Jenkins, test this please.

**Comment by damencho** [comment]
Feeback and minor improvements in the user experience. 
- Disable "Send" button if the poll cannot be send. Ideally to show what's missing ... but we can proceed with just the disabled button for now.
- The remove option to be grey, not red or better add it as a button circle with a X in it or something like that ... 
- The use case: you in a meeting and someone sends a poll, you see the badge that there is something new in the chat and when you open the chat you don't see what's new. We need some UI change to show new data in the Polls tab. The ideal is to have badge on the Polls tab.
- In regards to the previous one. If you have opened the Chat and Polls panel and you are focused on the Polls tab and someone in the meeting sends a message, there is no way to understand that there is a new message other than changing to the Chat tab. The ideal again is to have a badge in the Chat tab.
- We need to add a config.js config to be able to disable the polls, but by default, it is enabled and available.

Please ping me if you have any questions. 
Overall awesome work! @horymury is reviewing the UI code and will post any comments if there are any. I already went through it and haven't noticed anything. 
Thank you again!

File: react/features/polls/middleware.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    // Middleware triggered when a poll is received
+    case RECEIVE_POLL: {
+
+        const state = getState();
+        const isChatOpen: boolean = state['features/chat'].isOpen;
+        const isPollsTabFocused: boolean = state['features/chat'].isPollsTabFocused;
+
+        // Finally, we notify user they received a new poll if their pane is not opened
+        if (action.notify && (!isChatOpen || !isPollsTabFocused)) {
+            dispatch(playSound(INCOMING_MSG_SOUND_ID));
```
Comment (horymury) [inline react/features/polls/middleware.js:24]: to be discussed: maybe have a different sound for polls.

File: react/features/polls/subscriber.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        lastVote: null,
+        answers: answers2
+    };
+};
+
+StateListenerRegistry.register(
+    state => getCurrentConference(state),
+    (conference, store, previousConference) => {
+        if (conference && conference !== previousConference) {
+            conference.room.addListener('xmmp.json_message_received', (senderJid, data) => {
```
Comment (horymury) [inline react/features/polls/subscriber.js:59]: possible typo: `xmmp.`

File: react/features/polls/subscriber.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                if (data.type === COMMAND_NEW_POLL) {
```
Comment (horymury) [inline react/features/polls/subscriber.js:60]: nit: use switch

File: react/features/polls/subscriber.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        return null;
+    }
+    const { id, senderId, senderName, question, answers } = pollData;
+
+    if (typeof id !== 'string' || typeof senderId !== 'string' || typeof senderName !== 'string'
+        || typeof question !== 'string' || !(answers instanceof Array)) {
+        return null;
+    }
+
+    const answers2 = [];
```
Comment (horymury) [inline react/features/polls/subscriber.js:27]: nit : `answersParsed` ?

**Review by horymury — CHANGES_REQUESTED** [review]

**Comment by horymury** [comment]
@jade-guiton the code looks very neat and clean, really nice job! added just a few nitpicks.

Also added a suggestion, should we have different sound for polls notifications vs chat notifs? @damencho  WDYT?

File: react/features/polls/subscriber.js (hunk 2 above)
Comment (jade-guiton) [reply react/features/polls/subscriber.js:59]: That constant does not exist. The JSON_MESSAGE_RECEIVED event is internal to LJM, so the constant associated to it is not exported.

More importantly, now that [this PR on LJM](https://github.com/jitsi/lib-jitsi-meet/pull/1671) has been merged, this line will have to change anyway to use the new events (see [this PR on our fork](https://github.com/jade-guiton/jitsi-meet/pull/27)).

Comment (jade-guiton) [reply react/features/polls/subscriber.js]: The change in question was just merged into the PR branch. You can look at 5fc106b to see what changed regarding this issue.

**Comment by damencho** [comment]
> @jade-guiton the code looks very neat and clean, really nice job! added just a few nitpicks.
> 
> Also added a suggestion, should we have different sound for polls notifications vs chat notifs? @damencho WDYT?

I don't think that's a blocker even if you want to have different sound. But the badges I mentioned in my previous comment will solve the user experience and we don't need a different sound, in my opinion. But we can check this woth Emil at some point.

**Comment by fabienzucchet** [comment]
> @jade-guiton the code looks very neat and clean, really nice job! added just a few nitpicks.
> 
> Also added a suggestion, should we have different sound for polls notifications vs chat notifs? @damencho WDYT?

@horymury Thanks for your feedback ! We applied your nitpicks except for the comment about a different sound for polls as it is still under discussion

File: react/features/chat/actions.any.js [hunk 5]
```diff
@@ -97,3 +98,16 @@ export function setPrivateMessageRecipient(participant: Object) {
...
     };
 }
+
+/**
+ * Set the value of _isPollsTabFocused.
+ *
+ * @param {boolean} isPollsTabFocused - The new value for _isPollsTabFocused.
+ * @returns {Function}
+ */
+export function setisPollsTabFocused(isPollsTabFocused: boolean) {
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/actions.any.js]: Fixed

File: react/features/chat/components/native/Chat.js [hunk 6]
```diff
@@ -46,9 +50,32 @@ class Chat extends AbstractChat<Props> {
                 modalId = { CHAT_VIEW_MODAL_ID }
                 onClose = { this._onClose }>
 
-                <MessageContainer messages = { this.props._messages } />
-                <MessageRecipient />
-                <ChatInputBar onSend = { this._onSendMessage } />
+                <View style = { styles.tabContainer } >
+                    <Button
+                        color = '#17a0db'
+                        mode = { this.props._isPollsTabFocused ? 'contained' : 'text' }
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/components/native/Chat.js]: Fixed

File: react/features/chat/components/native/ChatMessage.js [hunk 7]
```diff
@@ -154,6 +154,7 @@ class ChatMessage extends AbstractChatMessage<Props> {
         );
     }
 
+
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/components/native/ChatMessage.js]: Fixed

File: react/features/chat/components/web/Chat.js [hunk 8]
```diff
@@ -19,6 +20,7 @@ import MessageContainer from './MessageContainer';
 import MessageRecipient from './MessageRecipient';
 import TouchmoveHack from './TouchmoveHack';
 
+
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/components/web/Chat.js]: Fixed

File: react/features/chat/components/web/Chat.js [hunk 9]
```diff
@@ -160,6 +198,7 @@ class Chat extends AbstractChat<Props> {
         );
     }
 
+
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/components/web/Chat.js]: Fixed

File: react/features/polls/components/AbstractPollAnswer.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+ */
+const AbstractPollAnswer = (Component: AbstractComponent<AbstractProps>) => (props: InputProps) => {
+
+    const { pollId } = props;
+
+    const conference: Object = useSelector(state => state['features/base/conference'].conference);
+
+    const poll: Poll = useSelector(state => state['features/polls'].polls[pollId]);
+
+    const localId: string = useSelector(state => getLocalParticipant(state).id);
```
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/polls/components/AbstractPollAnswer.js]: It works indeed ! Fixed

File: react/features/polls/subscriber.js (hunk 4 above)
Comment (fabienzucchet) [reply react/features/polls/subscriber.js]: Fixed

File: react/features/chat/components/native/Chat.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (horymury) [reply react/features/chat/components/native/Chat.js]: I think you missed a few places where it's still the hardcoded value :)

**Comment by horymury** [comment]
> > @jade-guiton the code looks very neat and clean, really nice job! added just a few nitpicks.
> > Also added a suggestion, should we have different sound for polls notifications vs chat notifs? @damencho WDYT?
> 
> @horymury Thanks for your feedback ! We applied your nitpicks except for the comment about a different sound for polls as it is still under discussion

Thank you for addressing them. 

Asked Emil about having different sound for polls notifications, he said that it's needed but that it should not block this PR.

File: react/features/chat/components/native/Chat.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (fabienzucchet) [reply react/features/chat/components/native/Chat.js]: Sorry I missed it I will fix it asap :/

**Comment by damencho** [comment]
Jenkins, test this please.

**Review by damencho — APPROVED** [review]

**Comment by damencho** [comment]
Jenkins, test this please.

**Comment by damencho** [comment]
Just waiting for @horymury to approve it as github don't allow me to merge otherwise (he told me offline everything is good).

Thanks for the quick fixes and addressing the review.

**Review by horymury — APPROVED** [review]

**Comment by damencho** [comment]
Tadaaa 🎉🎉🎉🎉🎉 :) 
Thank you all!

**Comment by saghul** [comment]
Amazing stuff folks! 🙌🙌👏👏👏

**Comment by sampaccoud** [comment]
Nice one bravo ! 👏👏👏

**Comment by pierreozoux** [comment]
Wow :) Nice to see this merged :)

Thank you everybody!

**Comment by Marinaarimany** [comment]
Hello, can someone add the deactivation of polls in the feature flags for mobile please ?

https://github.com/jitsi/jitsi-meet/issues/10033

**Comment by Marinaarimany** [comment]
@jade-guiton :)

**Comment by Marinaarimany** [comment]
> Hello, can someone add the deactivation of polls in the feature flags for mobile please ?
> 
> #10033

@jade-guiton :)

**Comment by saghul** [comment]
@Marinaarimany please stop pestering people.

**Comment by Marinaarimany** [comment]
> @Marinaarimany please stop pestering people.

oups sorry :/

**Comment by saghul** [comment]
"Over complicating" here is about the Ui/UX and implementation, not necessarily the concept.

While polls might not be perfect, we think it's at the right spot at the moment.

**Comment by Bodobolero** [comment]
Above you said: "there are plenty of services that do exactly" - can you link some open source and/or free tools that do that exactly? I researched and could only find commercial (not cheap) services.

**Comment by keunes** [comment]
@Bodobolero It's out of scope of this issue really, but you might have a look at https://framadate.org (the name is a bit misleading - it doesn't only do dates). Don't know if it has all the features you're looking for, though.

**Comment by Bodobolero** [comment]
@keunes Thank you for the link. I tried https://framadate.org 
It has the same limitations as Jitsi poll - it does not allow to create a question with single-choice. It only allows multiple choice questions. So you can not use it to vote - e.g. for a person.

**Pull request merged**